accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,547 @@
+- Type
+  - Value types
+    - Built-in value types
+    - User-defined value types
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // These throw InvalidOperationException if IsFunctionPointer is false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // These only work for types obtained from GetSignature*Type().
+        // An empty Type[] is returned otherwise.
++       public virtual Type[] GetFunctionPointerCallingConventions();
++       public Type[] GetRequiredCustomModifiers();
++       public Type[] GetOptionalCustomModifiers();
+    }
+}
+```
+### Examples
+```cs
+// Simple managed function pointer
+Type type = typeof(delegate*<int, bool>);
+Debug.Assert(type.IsFunctionPointer == true);
+Debug.Assert(type.IsUnmanagedFunctionPointer == false);
+Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
+Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
+Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
+Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 0);
+```
+
+```cs
+// Simple unmanaged function pointer
+Type type = typeof(delegate* unmanaged[Cdecl, MemberFunction]<int, bool>);
+Debug.Assert(type.IsFunctionPointer == true);
+Debug.Assert(type.IsUnmanagedFunctionPointer == true);
+Debug.Assert(type.GetFunctionPointerReturnType() == typeof(bool));
+
+Debug.Assert(type.GetFunctionPointerParameterTypes().Length == 1);
+Debug.Assert(type.GetFunctionPointerParameterTypes()[0] == typeof(int));
+
+Debug.Assert(type.GetFunctionPointerCallingConventions().Length == 2);
+Debug.Assert(type.GetFunctionPointerCallingConventions[0] == typeof(CallConvCdecl));
+Debug.Assert(type.GetFunctionPointerCallingConventions[1] == typeof(CallConvMemberFunction));
+```
+
+## GetSignature*Type() methods for `FieldInfo`, `PropertyInfo` and `ParameterInfo`
+These return a function pointer `Type` that has references to custom modifiers.
+
+If not a function pointer, these methods return the value from the appropriate `FieldInfo.FieldType`, `PropertyInfo.PropertyType` or `ParameterInfo.ParameterType` property.
+
+```diff
+namespace System.Reflection
+{
+    public class FieldInfo
+    {
++       public Type GetSignatureFieldType();
+    }
+
+    public class PropertyInfo
+    {
++       public Type GetSignaturePropertyType();
+    }
+
+    public class ParameterInfo
+    {
++       public Type GetSignatureParameterType();
+    }
```

Comment:
```suggestion
    public abstract class FieldInfo
    {
+       public virtual Type GetSignatureFieldType();
    }

    public abstract class PropertyInfo
    {
+       public virtual Type GetSignaturePropertyType();
    }

    public abstract class ParameterInfo
    {
+       public virtual Type GetSignatureParameterType();
    }
```

Reply:
Also, what is going to be the default implementation of these methods?

Reply:
The default implementation will just return `FieldInfo.FieldType`, etc. I don't think we want to throw if a non function pointer.

Reply:
Update: I changed these to return `null` if the FieldType etc is not a function pointer. Also related is that a modified type is just a "wrapper" on top of the underlying type that doesn't pretend to be the actual underlying type.